openid/openid.go

```diff
@@ -0,0 +1,242 @@
+		"openid.return_to":  {r.ReturnTo},
+	}
+	if r.Realm != "" {
+		v.Set("openid.realm", r.Realm)
+	}
+	if len(r.Teams) > 0 {
+		v.Set("openid.ns.lp", nsTeams)
+		v.Set("openid.lp.query_membership", strings.Join(r.Teams, ","))
+	}
+	if len(r.SimpleRegistrationRequired) > 0 {
+		v.Set("openid.ns.sreg", nsSimpleRegistration)
+		v.Set("openid.sreg.required", strings.Join(r.SimpleRegistrationRequired, ","))
+	}
+	if len(r.SimpleRegistrationOptional) > 0 {
+		v.Set("openid.ns.sreg", nsSimpleRegistration)
+		v.Set("openid.sreg.optional", strings.Join(r.SimpleRegistrationOptional, ","))
+	}
+	return server.OpenIDURL() + "?" + v.Encode()
+}
+
+// Respones contains the values returned from Ubuntu SSO after a
+// successful login.
+type Response struct {
+	// ID contins the claimed_id of the logged in user. This will
+	// always be present in a successful login.
+	ID string
+
+	// Teams contains any launchpad teams, that were queried in the
+	// request, that the logged in user is a member of.
+	Teams []string
+
+	// SimpleRegistration contains any simple registration fileds are
+	// were assered in the response.
+	SimpleRegistration map[string]string
+}
+
+// verify is used to perform the OpenID verification of the login response.
+var verify = openid.Verify
+
+// Verify processes a positive assertion from Ubuntu SSO. If the
+// verification is successful any parameters asserted by Ubuntu SOO will
+// be set in the Response.
+func Verify(requestURL string, server usso.UbuntuSSOServer, ns NonceStore) (*Response, error) {
+	u, err := url.ParseRequestURI(requestURL)
+	if err != nil {
+		return nil, err
+	}
+	v := u.Query()
+	switch v.Get("openid.mode") {
+	case "error":
+		return nil, fmt.Errorf("%s", v.Get("openid.error"))
+	case "cancel":
+		return nil, fmt.Errorf("login cancelled")
+	default:
+		return nil, fmt.Errorf("unrecognised mode %q", v.Get("openid.mode"))
+	case "id_res":
+	}
+	if endpoint := v.Get("openid.op_endpoint"); endpoint != server.OpenIDURL() {
+		return nil, fmt.Errorf("OpenID response from unexpected endpoint %q", endpoint)
+	}
+	// verify the openid response.
+	if ns == nil {
+		ns = DefaultNonceStore
+	}
+	dc := DefaultDiscoveryCache
+	if dc == nil {
+		dc = ussoDiscoveryCache(server)
+	}
+	id, err := verify(requestURL, dc, ns)
+	if err != nil {
+		return nil, err
+	}
+	r := Response{
+		ID: id,
+	}
+	// check for extensions in the response.
+	signed := strings.Split(v.Get("openid.signed"), ",")
+	if v.Get("openid.ns.lp") == nsTeams && contains("lp.is_member", signed) {
+		r.Teams = strings.Split(v.Get("openid.lp.is_member"), ",")
+	}
+	if v.Get("openid.ns.sreg") == nsSimpleRegistration {
+		r.SimpleRegistration = make(map[string]string)
+		for k := range v {
+			if !strings.HasPrefix(k, "openid.sreg.") {
+				continue
+			}
+			if !contains(k[7:], signed) {
+				continue
+			}
+			r.SimpleRegistration[k[12:]] = v.Get(k)
+		}
+		if len(r.SimpleRegistration) == 0 {
```

Comment:
Rather than this, you could move the if statement just before
the map assignment:

     if r.SimpleRegistration == nil {
          r.SimpleRegistration = make(...)
     }

then we'd save making a map and throwing it away again.
*shrug*